server/sonar-web/src/main/js/apps/custom-measures/components/Form.tsx

```diff
@@ -0,0 +1,209 @@
+/*
+ * SonarQube
+ * Copyright (C) 2009-2018 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+import * as React from 'react';
+import { getAllMetrics } from '../../../api/metrics';
+import { CustomMeasure, Metric } from '../../../app/types';
+import DeferredSpinner from '../../../components/common/DeferredSpinner';
+import Select from '../../../components/controls/Select';
+import SimpleModal from '../../../components/controls/SimpleModal';
+import { translate } from '../../../helpers/l10n';
+
+interface Props {
+  confirmButtonText: string;
+  measure?: CustomMeasure;
+  header: string;
+  onClose: () => void;
+  onSubmit: (data: { description: string; metricKey: string; value: string }) => Promise<void>;
+  skipMetrics?: string[];
+}
+
+interface State {
+  description: string;
+  loading: boolean;
+  metricKey?: string;
+  metrics?: Metric[];
+  value: string;
+}
+
+export default class Form extends React.PureComponent<Props, State> {
+  mounted: boolean;
+
+  constructor(props: Props) {
+    super(props);
+    this.state = {
+      description: (props.measure && props.measure.description) || '',
+      loading: false,
+      metricKey: props.measure && props.measure.metric.key,
+      value: (props.measure && props.measure.value) || ''
+    };
+  }
+
+  componentDidMount() {
+    this.mounted = true;
+    if (!this.props.measure) {
+      this.fetchCustomMetrics();
+    }
+  }
+
+  componentWillUnmount() {
+    this.mounted = false;
+  }
+
+  handleSubmit = () => {
+    return this.state.metricKey
+      ? this.props
+          .onSubmit({
+            description: this.state.description,
+            metricKey: this.state.metricKey,
+            value: this.state.value
+          })
+          .then(this.props.onClose)
+      : Promise.reject(undefined);
+  };
+
+  fetchCustomMetrics = () => {
+    this.setState({ loading: true });
+    getAllMetrics({ isCustom: true }).then(
```

Comment:
Maybe you could use `getMetrics` from the `rootReducer` if they are already available and/or ensure they are available with `fetchMetrics` from `rootActions`.

Reply:
I'd need to pass an additional parameter `isCustom`. Also I'd need to handle loading of metrics differently. Having these complexities in mind, I'd not change my implementation :)

Reply:
Ah yeah, I didn't think about that `isCustom` flag